reana_client/api/client.py

```diff
@@ -290,21 +294,24 @@ def create_workflow_from_json(
         )
     try:
         reana_yaml = dict(workflow={})
-        if workflow_file:
-            reana_yaml["workflow"]["file"] = workflow_file
-        else:
-            reana_yaml["workflow"]["specification"] = workflow_json
         reana_yaml["workflow"]["type"] = workflow_engine
         if parameters:
             reana_yaml["inputs"] = parameters
         if outputs:
             reana_yaml["outputs"] = outputs
+        if workflow_file:
+            reana_yaml["workflow"]["file"] = workflow_file
+            reana_yaml["workflow"][
+                "specification"
+            ] = load_workflow_spec_from_reana_yaml(reana_yaml)
+            input_params = load_input_parameters(reana_yaml)
+            if input_params:
+                reana_yaml["inputs"]["parameters"] = input_params
```

Comment:
I guess this part should go inside the `if parameters:` section?

Reply:
TL;DR I think that putting this part in the `if parameters:` section would make the validation of the Snakemake specification fail, because the parameters would be loaded too soon and replace the config files needed to validate the snakefile.

`reana_yaml` before `load_input_parameters` (pointing to the yaml file with the configuration parameters):
```yaml
inputs:
   parameters:
       input: workflow/Snakemake/helloworld-parameters.yaml
```

And `reana_yaml` after `load_input_parameters` (indeed loading the parameters contained in the yaml file):
```yaml
inputs:
   parameters:
      helloworld: code/helloworld.py
      inputfile: data/names.txt
      sleeptime: 0
```

In the Snakefile, the parameters are taken directly from the configfile:
```python
rule helloworld:
    input:
        helloworld=config["helloworld"],
        inputfile=config["inputfile"],
```

`load_workflow_spec_from_reana_yaml` loads the Snakemake / CWL / yadage specification and calls the respective validation functions. When `r-commons` [calls the snakemake validation function](https://github.com/reanahub/reana-commons/blob/3729df242049fd73c067aa2dcac3f16c242bbd09/reana_commons/snakemake.py#L172), it expects the configfile to be [provided in the kwargs](https://github.com/reanahub/reana-commons/blob/3729df242049fd73c067aa2dcac3f16c242bbd09/reana_commons/snakemake.py#L170). By the way the kwargs are prepared, the configfile is [taken from the `inputs.parameters.input` field](https://github.com/reanahub/reana-commons/pull/413/files#diff-4f3e2c0b81229cc70894a2058f2af7202cb2d2870624680715a09c3ea7c69e05R82-R85), thus this field should not be replaced by the actual parameters yet: that's why I put `load_input_parameters` after `load_workflow_spec_from_reana_yaml`.

It's not a very straightforward process, and the different types of "inputs" don't make things easier, but I hope this was not too hard to follow. What do you think?